### ISSUE TITLE
Apply dim reduction on the whole training Set?

### ISSUE THREAD

**Issue by charx7** [RUG-IDS/Course-2018#57]
Hi Ta's, we are wondering if we should apply dim reduction techniques such as the anova and the pca with the whole training dataset or should we apply them after each iteration of k-folds ie, after reserving one part of the training dataset for the final test?

Also, should we apply them before or after the upsampling/undersampling methods?

thanks!

**Comment by charx7** [comment]
Okay here is our pipeline,

[Get Data Clean] -> [Scale] -> [applyPCA/anova] -> [ upsample/undersample (class balance)] -> [ k-folds/X-validation] -> [Model Evaluation]

or should it be: 

[Get Data Clean] -> [Scale] ->  [ upsample/undersample (class balance)] -> [applyPCA/anova] -> [ k-folds/X-validation] -> [Model Evaluation]

thanks!

**Comment by rm-bwana** [comment]
Again. It seems you are trying to combine 2 or more questions into one. Using your structure I would say:

For Q1.1:
[Get Data Clean] -> [Scale] -> [applyPCA/anova] -> Learn what features are important and how the data is spread to help know what features to take forward.

But even here you should try out different preprocessing steps and then test with tSNE how different it looks. No k-folds required at this stage. 

For Q1.2
[Get Data Clean] -> [Scale] -> [ upsample/undersample (class balance)] -> [ k-folds/X-validation] -> [Model Evaluation]

PLEASE NOTE:  ANOVA helps in describing your data. It is not a preprocessing step.
Scaling and sampling are preprocessing steps so they should be present. PCA can be used as part of preprocessing but it depends on the algorithm to be used. 

If I have not understood you or if you are asking about a specific question then please say so. 

Rob.

**Comment by charx7** [comment]
Is for question 1.2 During the baseline experiments, we are using anova and/or PCA as a dimentionallity reduction technique to select the most promising features and then have model tryouts, to see what is the effect of dimentionallity reduction on the test sets. The thing is that we arent so sure if upsampling/undersampling 'bleeds' information to the test set of the model. 

As I understand you the features for question 1.2 should be the features that were selected by the analysis on 1.1
Thanks.

**Comment by rm-bwana** [comment]
Hi,

What do you refer to as 'bleeding' information?

In terms of train/test data if you are worried that the sampling technique will cause copies of data to be in both train and test then extract the test data first (which should include enough copies of both classes) and then up/down sample from the remaining to get your train data.

You then carry out experiments on the train data and use the test data to evaluate. If you have done preprocessing on your train data you must also apply it to the test data but independently (i.e. you cannot do it as one big batch). If there are certain preprocessing values used on the train data e.g. mean value, you have to use the exact same value for the test set (i.e. you subtract the train data mean from the test set, not the test data mean) and so on.

Rob.

**Comment by charx7** [comment]
Yes I was worried that the test set would not be an accurate representation to test the model due to the artificial samples created by upsampling, but since we dont have enough data to do many splits it would be convenient just to do 1 train/test split on the upsampled data.

Thanks for the answers!
